playbooks/common/openshift-master/config.yml

```diff
@@ -337,12 +361,54 @@
     with_items: openshift_master_named_certificates
     when: named_certs_specified | bool
 
-- name: Configure master instances
-  hosts: oo_masters_to_config
+- name: Configure first master instance
```

Comment:
Yeah it's presently the playbook/orchestrator's role to determine openshift_version from inventory vars, or protect the installed version by passing in the current version from pre-existing fact. You've hit upon the reason why that is, we can't have the role protect the installed version fact or we'd need an upgrade flag, which we're already trying to avoid.

Is there another option? 

What if I consider openshift_version an override instead of something that's almost always set, and push more of the logic into openshift_docker role.

- config playbook for first master passes no openshift_version
- openshift_docker role inherits responsibility of sorting out the various inventory vars and setting openshift_version:
  - if openshift_version passed use it (still converts generic 3.2 to specific 3.2.0.44 if necessary)
  - else use installed version if present
  - else use inventory vars if present (release / pkg_version / image_tag)
    - only openshift_docker will reference these, everything else uses openshift_version
  - else use lookup latest and use that
- config playbook for subsequent masters and nodes will pass openshift_version = first master installed version fact
- upgrade playbook will need to determine always set openshift_version.

Now the role defaults to protecting installed version, and installing latest if nothing is requested. It involves no upgrade flag, but orchestrator can request a specific version that takes over everything for upgrades and/or pinning to master version.

We do have to choose one of (a) protect installed versions and (b) sync everything to first master. Which one will customers prefer? 

I believe the above can solve both cases, we'd just have the config playbook *also* check for pre-existing version and pass it as openshift_version, otherwise pass first master.